core/src/crds_gossip_push.rs

```diff
@@ -67,10 +69,70 @@ impl CrdsGossipPush {
     pub fn num_pending(&self) -> usize {
         self.push_messages.len()
     }
+
+    fn prune_stake_threshold(self_stake: u64, origin_stake: u64) -> u64 {
+        let min_path_stake = self_stake.min(origin_stake);
+        (CRDS_GOSSIP_PRUNE_STAKE_THRESHOLD_PCT * min_path_stake as f64).round() as u64
+    }
+
+    pub fn prune_received_cache(
+        &mut self,
+        self_pubkey: &Pubkey,
+        origin: &Pubkey,
+        hash: Hash,
+        stakes: &HashMap<Pubkey, u64>,
+    ) -> Vec<Pubkey> {
+        let origin_stake = stakes.get(origin);
+        let self_stake = stakes.get(self_pubkey);
+        let cache = self.received_cache.get(&hash);
+        if cache.is_none() || origin_stake.is_none() || self_stake.is_none() {
+            return Vec::new();
+        }
+
+        let peers = &cache.unwrap().1;
+        let staked_peers: Vec<(Pubkey, u64)> = peers
+            .iter()
+            .filter_map(|p| stakes.get(p).map(|s| (*p, *s)))
+            .filter(|(_, s)| *s > 0)
+            .collect();
+
+        let peer_stake_total: u64 = staked_peers.iter().map(|(_, s)| s).sum();
```

Comment:
nit: you can avoid constructing staked_peers if prune_stake_threshold isn't met by summing the stakes before filtering by stake != 0..